ec-gpu/src/lib.rs

```diff
@@ -1,22 +1,43 @@
-/// Describes how to generate the elliptic curve operations for
-/// - `Scalar`
-/// - `Fp`
-/// - `Fp2`
-/// - `G1`
-/// - `G2`
-pub trait GpuEngine {
-    type Scalar: GpuField;
-    type Fp: GpuField;
+/// The name that is used in the GPU source code to identify the item that is used.
+///
+/// It is automatically implemented for the [`GpuField`] trait.
+pub trait GpuName {
+    /// A unique name for the item.
+    ///
+    /// The unique identifier is not stable, i.e. different versions of Rust or this crate might
+    /// lead to different names. This is OK as kernel sources/binaries are always bundled with a
+    /// library and not re-used between versions.
+    ///
+    /// The default implementation is using the type itself together with the full module path. All
+    /// non-alphanumeric characters are replaced with underscores, so that it's an identifier that
+    /// doesn't cause any issues with C compilers. You should never need your own custom
+    /// implementation.
+    fn name() -> String {
+        //sanitize_name(&format!("{}_{}", module_path!(), std::any::type_name::<Self>()))
+        format!("{}_{}", module_path!(), std::any::type_name::<Self>())
+            .replace(|c: char| !c.is_ascii_alphanumeric(), "_")
+    }
```

Comment:
This default implementation doesn't quite do the right thing. `module_path!()` inserts the path for the module in which the macro is evaluated, which in this case is `ec_gpu`. That's handy for domain-separating multiple versions of `ec-gpu`, but this could also just be done at the point where `GpuName::name()` gets called.

What I'd suggested it for was domain-separating the downstream types themselves, and that only works if the `module_path!()` macro is invoked inside the downstream modules. This is why in my previous comment I suggested providing a wrapping macro that downstreams could use to implement `GpuName`, which _would_ invoke it in the correct context.

The reason the examples you showed before look right is because `std::any::type_name` happens to be printing the module path, but it is not guaranteed to do so (it may also just show the type name). I think the best way to combine them would be to check if the output of `type_name()` is prefixed by `module_path!()` and prepend the latter if not. In reusable macro form:
```rust
#[macro_export]
macro_rules! name {
    ($named:ident) => {
        impl ::ec_gpu::GpuName for $named {
            fn name() -> String {
                let mod_path = module_path!();
                let type_name = std::any::type_name::<Self>();
                if type_name.starts_with(mod_path) {
                    type_name.replace(|c: char| !c.is_ascii_alphanumeric(), "_")
                } else {
                    format!("{}__{}", mod_path, type_name)
                        .replace(|c: char| !c.is_ascii_alphanumeric(), "_")
                }
            }
        }
    };
}
```
and then the default impl would be removed:
```suggestion
    fn name() -> String;
```

Reply:
The macro would then be used like so:
```rust
#[cfg(feature = "gpu")]
ec_gpu::name!(Fp2);

#[cfg(feature = "gpu")]
impl ec_gpu::GpuField for Fp2 {
    ...
}
```